tuf/api/metadata.py

```diff
@@ -861,22 +904,84 @@ def to_dict(self) -> Dict[str, Any]:
         }
 
 
+class TargetFile:
+    """A container with information about a particular target file.
+
+    Attributes:
+        length: An integer indicating the length of the target file.
+        hashes: A dictionary mapping hash algorithms to the
+            hashes resulting from applying them over the metadata file
+            contents::
+
+              'hashes': {
+                    '<HASH ALGO 1>': '<TARGET FILE HASH 1>',
+                    '<HASH ALGO 2>': '<TARGET FILE HASH 2>',
+                    ...
+                }
+        custom: An optional dictionary which may include version numbers,
+            dependencies, or any other data that the application wants
+            to include to describe the target file::
+                'custom': {
+                    'type': 'metadata',
+                    'file_permissions': '0644',
+                    ...
+                } // optional
+
+        unrecognized_fields: Dictionary of all unrecognized fields.
+
+    """
+
+    def __init__(
+        self,
+        length: int,
+        hashes: Dict[str, str],
+        custom: Optional[Dict[str, Any]] = None,
```

Comment:
I like the idea of using custom as `unrecognized fields` through using a property. 
Will do it and will add a comment that custom will replace `unrecognized fields` for this particular case.

Reply:
Wondering, if we use `custom` replacing the need of `unrecognized fields` for this class, then do you test for this in #1382?
Do you add an unrecognized field under the `custom` dictionary?

Reply:
I was trying to suggest dropping "custom", and just letting the already existing unrecognised_fields functionality handle it -- that basically handles de/serializing both "custom" and any other unrecognised fields.

Then I was saying that we might still want to add a property "custom" (to conform with the API design principle that the object would copy the file format) but the property would just access `self.unrecognised_fields.get("custom")`.

What you implemented is close but not quite the same... now TargetFile cannot have unrecognised_fields at all (it can only have a "custom" field): as you correctly point out that means deciding that TargetFile is not allowed to have unrecognised fields and updating the test... which seems like the implementation detail guiding how we read the specification. 

I don't think this is a major issue, but I also think what I suggested would cover all the cases without any more complexity... opinions?

Reply:
Well, `custom` basically is the same as `unrecognised_fields`.
I feel like we should encourage our users to use `custom` as described in the spec when adding additional fields instead of relying on `unrecognized_fields` at least in this scenario.
Maybe we can achieve that by adding a `custom` property, but then if we want to treat it as a regular attribute maybe we would have to add a property setter as well?

On another hand, if we don't have `unrecognized_fields` in TargetFile, target.json file with unrecognized fields outside custom won't work...

Reply:
> Well, `custom` basically is the same as `unrecognised_fields`.
> I feel like we should encourage our users to use `custom` as described in the spec when adding additional fields instead of relying on `unrecognized_fields` at least in this scenario.
> Maybe we can achieve that by adding a `custom` property, but then if we want to treat it as a regular attribute maybe we would have to add a property setter as well?
> 
> On another hand, if we don't have `unrecognized_fields` in TargetFile, target.json file with unrecognized fields outside custom won't work...

that's what I'm saying: if we read the spec as "you need to be able to add new fields to any dictionary where it's technically possible" (which was the consensus I believe), then  implementing custom is not enough for spec-compliance

I don't think it's a major thing but ... if we can easily achieve spec compliance in a trivial matter then why not do it?

Reply:
Okay, will update the class, so that `unrecognized fields` is documented and added as an attribute, but I am not sure should we document `custom` when `custom` will be just a property and we retrieve with `self.unrecognised_fields.get("custom")`?

Reply:
I didn't define it in my last update. Do you think that's okay?